go/vt/vttablet/tabletmanager/vreplication/journal_test.go

```diff
@@ -0,0 +1,303 @@
+/*
+Copyright 2019 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package vreplication
+
+import (
+	"fmt"
+	"testing"
+
+	"golang.org/x/net/context"
+	binlogdatapb "vitess.io/vitess/go/vt/proto/binlogdata"
+)
+
+func TestJournalOneToOne(t *testing.T) {
+	defer deleteTablet(addTablet(100))
+	defer deleteTablet(addOtherTablet(101, "other_keyspace", "0"))
+
+	execStatements(t, []string{
+		"create table t(id int, val varbinary(128), primary key(id))",
+		fmt.Sprintf("create table %s.t(id int, val varbinary(128), primary key(id))", vrepldb),
+	})
+	defer execStatements(t, []string{
+		"drop table t",
+		fmt.Sprintf("drop table %s.t", vrepldb),
+	})
+	env.SchemaEngine.Reload(context.Background())
+
+	filter := &binlogdatapb.Filter{
+		Rules: []*binlogdatapb.Rule{{
+			Match: "t",
+		}},
+	}
+	_, firstID := startVReplication(t, filter, binlogdatapb.OnDDLAction_IGNORE, "")
+
+	journal := &binlogdatapb.Journal{
+		Id:            1,
+		MigrationType: binlogdatapb.MigrationType_SHARDS,
+		Participants: []*binlogdatapb.KeyspaceShard{{
+			Keyspace: "vttest",
+			Shard:    "0",
+		}},
+		ShardGtids: []*binlogdatapb.ShardGtid{{
+			Keyspace: "other_keyspace",
+			Shard:    "0",
+			Gtid:     "MySQL56/7b04699f-f5e9-11e9-bf88-9cb6d089e1c3:1-10",
+		}},
+	}
+	query := fmt.Sprintf("insert into _vt.resharding_journal(id, db_name, val) values (1, 'vttest', %v)", encodeString(journal.String()))
+	execStatements(t, []string{createReshardingJournalTable, query})
+	defer execStatements(t, []string{"delete from _vt.resharding_journal"})
+
+	expectDBClientQueries(t, []string{
+		"/update _vt.vreplication set pos=",
+		"begin",
+		`/insert into _vt.vreplication.*workflow, source, pos.*values.*'test', 'keyspace:\\"other_keyspace\\" shard:\\"0\\.*'MySQL56/7b04699f-f5e9-11e9-bf88-9cb6d089e1c3:1-10'`,
+		fmt.Sprintf("delete from _vt.vreplication where id=%d", firstID),
+		"commit",
+		"/update _vt.vreplication set state='Running', message='' where id.*",
+	})
+
+	// Delete all vreplication streams. There should be only one, but we don't know its id.
+	if _, err := playerEngine.Exec("delete from _vt.vreplication"); err != nil {
+		t.Fatal(err)
+	}
+	expectDeleteQueries(t)
+}
+
+func TestJournalOneToMany(t *testing.T) {
```

Comment:
Is there a many to one case that needs to be tested?

Reply:
I realized why I didn't write when I tried writing one. The current framework is not setup to handle this use case, because everything is in one mysql server. We'll need a real end to end test for this. I've added this to the TODO list so we don't forget.